test/test_multihead_attention_layer.f90

```diff
@@ -0,0 +1,412 @@
+program test_multihead_attention_layer
+  use iso_fortran_env, only: stderr => error_unit
+  use nf_multihead_attention_layer, only: multihead_attention_layer
+  use nf_self_attention_layer, only: self_attention_layer
+  use nf_cross_attention_layer, only: cross_attention_layer
+  use nf_linear2d_layer, only: linear2d_layer
+  use nf_optimizers, only: sgd
+  implicit none
+
+  logical :: ok = .true.
+  type(multihead_attention_layer) :: attention
+  real :: sample_input(3, 4) = reshape([0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 0.11, 0.12], [3, 4])
+  real :: split_heads_output(3, 2, 2)
+  real :: minput(3, 4) = reshape([0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 0.11, 0.12], [3, 4])
+  real :: output(3, 2, 2)
+
+  attention = multihead_attention_layer(n_heads=2)
+  call attention % init_base([3, 4])
+  call set_weights(attention)
+
+  call test_multihead_attention_split_heads(attention, sample_input, ok, split_heads_output)
+  call test_multihead_attention_create_attention_matrix(attention, split_heads_output, ok)
+  call test_multihead_attention_normalization(attention, ok)
+  call test_multihead_attention_scaled_dot_product_attention(attention, split_heads_output, ok)
+  call test_multihead_attention_combine_heads(attention, attention % sdpa, ok)
+  call test_multihead_attention_forward(attention, ok)
+  call test_multihead_attention_backward(attention, ok)
+  call test_multihead_attention_update_gradients(attention, ok)
+  call test_multihead_attention_forward_reallife_shape(ok)
+  call test_self_attention(ok)
+  call test_cross_attention(ok)
+
+  if (ok) then
+    print '(a)', 'test_multihead_attention_layer: All tests passed.'
+  else
+    write(stderr, '(a)') 'test_multihead_attention_layer: One or more tests failed.'
+    stop 1
+  end if
+
+contains
+  function allclose(x, y) result(res)
+    real, intent(in) :: x(:)
+    real, intent(in) :: y(:)
+    logical :: res
+
+    res = all(abs(x - y) <= (1e-06 + 1e-05 * abs(y)))
+  end function allclose
```

Comment:
Suggestion for future: create `nf_utils.f90` (or similar) and put this procedure there